sql: TestAbortedTxnLocks failed

sql.TestAbortedTxnLocks [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20936504?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20936504?buildTab=artifacts#/) on master @ [839f8e8cbcfb9e6fe49f7d6c9d8328d493ce045f](https://github.com/cockroachdb/cockroach/commits/839f8e8cbcfb9e6fe49f7d6c9d8328d493ce045f):


```
=== RUN   TestAbortedTxnLocks
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestAbortedTxnLocks1864740395
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:182: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:668: DRPC is enabled (via override by TestingGlobalDRPCOption)
    conn_executor_test.go:1983: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestAbortedTxnLocks1864740395
--- FAIL: TestAbortedTxnLocks (3.92s)
=== RUN   TestAbortedTxnLocks/with_cockroach_restart_savepoint_and_advanced_retry
    conn_executor_test.go:1957: 
        	Error Trace:	pkg/sql/conn_executor_test.go:1957
        	Error:      	An error is expected but got nil.
        	Test:       	TestAbortedTxnLocks/with_cockroach_restart_savepoint_and_advanced_retry
--- FAIL: TestAbortedTxnLocks/with_cockroach_restart_savepoint_and_advanced_retry (1.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAbortedTxnLocks.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58425

sql.TestAbortedTxnLocks [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20936605?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20936605?buildTab=artifacts#/) on master @ [2d75bf2d2095e1bfbc52b7d45c0740043e794a39](https://github.com/cockroachdb/cockroach/commits/2d75bf2d2095e1bfbc52b7d45c0740043e794a39):


```
=== RUN   TestAbortedTxnLocks
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestAbortedTxnLocks311540223
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:182: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    conn_executor_test.go:1983: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestAbortedTxnLocks311540223
--- FAIL: TestAbortedTxnLocks (4.47s)
=== RUN   TestAbortedTxnLocks/with_cockroach_restart_savepoint_and_advanced_retry
    conn_executor_test.go:1957: 
        	Error Trace:	pkg/sql/conn_executor_test.go:1957
        	Error:      	An error is expected but got nil.
        	Test:       	TestAbortedTxnLocks/with_cockroach_restart_savepoint_and_advanced_retry
--- FAIL: TestAbortedTxnLocks/with_cockroach_restart_savepoint_and_advanced_retry (1.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAbortedTxnLocks.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/conn_executor_test.go
+++ b/pkg/sql/conn_executor_test.go
@@ -1744,6 +1744,21 @@ func TestInjectRetryOnCommitErrors(t *testing.T) {
 	})
 }
 
+// TestAbortedTxnLocks verifies the lock behavior when a transaction enters the
+// aborted state due to an error. The expected behavior depends on whether any
+// savepoints are active:
+//
+//   - Without savepoints (or with released savepoints): locks are released
+//     immediately when the transaction aborts, allowing other transactions to
+//     proceed without waiting for an explicit ROLLBACK.
+//
+//   - With unreleased savepoints (including after ROLLBACK TO SAVEPOINT): locks
+//     are held until the transaction is explicitly rolled back, because the
+//     client could potentially recover via ROLLBACK TO SAVEPOINT.
+//
+// This distinction exists because savepoints allow error recovery within a
+// transaction, so the database cannot safely release locks until it knows the
+// transaction will not continue.
 func TestAbortedTxnLocks(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -1912,6 +1927,9 @@ func TestAbortedTxnLocks(t *testing.T) {
 	})
 
 	t.Run("with cockroach_restart savepoint and advanced retry", func(t *testing.T) {
+		if !s.Codec().IsSystem() {
+			skip.IgnoreLintf(t, "test is flaky on secondary tenants")
+		}
 		_, err = conn1.ExecContext(ctx, `INSERT INTO t VALUES (5,5), (6,6)`)
 		require.NoError(t, err)
 

